cobalt/browser/metrics/cpu_usage_tracker.cc

```diff
@@ -0,0 +1,230 @@
+          .Append(base::Value::Dict().Set("type", "total").Set("seconds", 30))
+          .Append(
+              base::Value::Dict().Set("type", "per_thread").Set("seconds", 2))
+          .Append(base::Value::Dict()
+                      .Set("type", "per_thread")
+                      .Set("seconds", 30)));
+}
+
+absl::optional<Config> ParseConfig(const base::Value& config_value) {
+  const base::Value::List* list = config_value.GetIfList();
+  if (!list) return absl::nullopt;
+  Config config;
+  for (auto& item_value : *list) {
+    const base::Value::Dict* item = item_value.GetIfDict();
+    if (!item) return absl::nullopt;
+    const std::string* type = item->FindString("type");
+    if (!type) return absl::nullopt;
+    Interval interval;
+    if (type->compare("total") == 0) {
+      interval.type = IntervalType::TOTAL;
+    } else if (type->compare("per_thread") == 0) {
+      interval.type = IntervalType::PER_THREAD;
+    } else {
+      return absl::nullopt;
+    }
+    interval.seconds = item->FindInt("seconds").value_or(0);
+    if (interval.seconds == 0) return absl::nullopt;
+    config.intervals.push_back(interval);
+  }
+  return std::move(config);
+}
+
+bool ValidateConfig(const base::Value& config) {
+  return ParseConfig(config).has_value();
+}
+
+}  // namespace
+
+// static
+CPUUsageTracker* CPUUsageTracker::GetInstance() {
+  return base::Singleton<CPUUsageTracker,
+                         base::LeakySingletonTraits<CPUUsageTracker>>::get();
+}
+
+CPUUsageTracker::CPUUsageTracker()
+    : storage_(nullptr),
+      task_runner_(base::SequencedTaskRunner::GetCurrentDefault()) {}
+
+void CPUUsageTracker::Initialize(
+    persistent_storage::PersistentSettings* storage) {
+  DCHECK_EQ(storage_, nullptr);
+  storage_ = storage;
+  task_runner_->PostTask(FROM_HERE,
+                         base::BindOnce(&CPUUsageTracker::InitializeAsync,
+                                        base::Unretained(this)));
+}
+
+void CPUUsageTracker::InitializeAsync() {
+  base::ProcessMetricsHelper::PopulateClockTicksPerS();
+  base::Value config;
+  storage_->Get("cpu_usage_tracker_intervals", &config);
+  UpdateConfig(config);
+}
+
+void CPUUsageTracker::UpdateConfig(const base::Value& config_value) {
+  if (!task_runner_->RunsTasksInCurrentSequence()) {
+    task_runner_->PostTask(
+        FROM_HERE,
+        base::BindOnce(&CPUUsageTracker::UpdateConfig, base::Unretained(this),
+                       config_value.Clone()));
+    return;
+  }
+  absl::optional<Config> config = ParseConfig(config_value);
+  if (!config.has_value()) {
+    base::Value default_config = GetDefaultConfigValue();
+    config = ParseConfig(default_config);
+    DCHECK(config.has_value());
+    storage_->Set("cpu_usage_tracker_intervals", std::move(default_config));
+  } else {
+    storage_->Set("cpu_usage_tracker_intervals", config_value.Clone());
+  }
+  base::AutoLock auto_per_thread_lock(per_thread_lock_);
```

Comment:
Nit: To improve readability, I think we can break this function into a few smaller ones, or maybe add some comments for each block?